### PR TITLE
feature/MessagesRedesign

### PR DESCRIPTION
**Pull request by joeykleingers**
* Replacing PipelineMessage API with an AbstractMessage superclass and subclasses for every type of message. This eliminates the need for the MessageType enumeration.

* Implementing a new, cleaner and more reusable API that observers can implement to be able to process specific types of messages.  Take a look at the IssuesWidget and IssuesWidgetMessageHandler for a good example of this.  Observers can decide which types of messages they are interested in by subclassing the AbstractMessageHandler class and reimplement the appropriate processMessage() methods.

* Passing AbstractMessages from observable object to observer object using a shared pointer instead of by value. This allows observers to store the messages with all the relevant subclass data if they prefer.

* Eliminating the use of humanLabel in Observable's and AbstractFilter's notify****Message method parameters. In AbstractFilter, these methods already set the human label into the message.

* It is no longer necessary to use setErrorCondition() or setWarningCondition() outside of the notify****Message methods. The notify methods will automatically set the condition to the code that is passed in.

* Filters that emit progress messages while they are in an executing pipeline have their progress values automatically converted into overall pipeline progress values and re-emitted as pipeline progress messages. This allows filters to update the overall progress of the pipeline while they are running. Check out FilterPipelineMessageHandler at the top of the FilterPipeline implementation file.

### CODE REVIEW

File: Source/DevHelper/Template/Filter/Filter.cpp.in [hunk 1]
```diff
@@ -83,15 +83,15 @@ void @ClassName@::execute()
   if (getWarningCondition() < 0)
   {
     QString ss = QObject::tr("Some warning message");
-    setWarningCondition(-88888888);
-    notifyWarningMessage(getHumanLabel(), ss, getWarningCondition());
+    setWarningCondition(-88888888); // @@WARNING_CODE@@-88888888@@
+    @@SET_WARNING_CONDITION@@(@@getWarningCondition()@@, ss);
```
Comment (mmarineBlueQuartz) [inline Source/DevHelper/Template/Filter/Filter.cpp.in]: `setWarningCondition(int)` was removed.

File: Source/DevHelper/Template/Filter/Filter.cpp.in [hunk 2]
```diff
@@ -83,15 +83,15 @@ void @ClassName@::execute()
...
   }
 
   if (getErrorCondition() < 0)
   {
     QString ss = QObject::tr("Some error message");
-    setErrorCondition(-99999999);
-    notifyErrorMessage(getHumanLabel(), ss, getErrorCondition());
+    setErrorCondition(-99999999); // @@ERROR_CODE@@-99999999@@
+    @@SET_ERROR_CONDITION@@(@@getErrorCondition()@@, ss);
```
Comment (mmarineBlueQuartz) [inline Source/DevHelper/Template/Filter/Filter.cpp.in]: `setErrorCondition(int)` was removed.

File: Source/DevHelper/Template/Filter/Filter.cpp.in (hunk 1 above)
Comment (joeykleingers) [reply Source/DevHelper/Template/Filter/Filter.cpp.in]: Ah yes.  My tool only hit .cpp and .h files and Qt Creator does not include those template files when I do a mass search to check that I replaced them all.  I will fix them.

**Review by mmarineBlueQuartz — APPROVED** [review]
Clang-format SIMPLViewUIMessageHandler.h.  Everything else looks good to me.

**Comment by joeykleingers** [comment]
I have a pre-commit hook on all my repos now that will clang-format it for me before committing, so a lot of these formatting issues should be cleaned up in future PRs.

**Pull request merged**